File: jebat_sdk/src/jebat_sdk/async_client.py
"""
Asynchronous JEBAT API Client.

Provides async interface to the JEBAT REST API with full streaming support.
"""

import asyncio
import json
from typing import Optional, Dict, Any, List, Union, AsyncIterator
from contextlib import asynccontextmanager

import httpx
from pydantic import BaseModel

from .models import *
from .exceptions import (
    JebatError,
    map_http_error,
    AuthenticationError,
    RateLimitError,
    ServerError,
)
from .retry import standard_retry, is_retryable_exception


class AsyncJebatClient:
    """
    Asynchronous JEBAT API client with full streaming support.

    Usage:
        async with AsyncJebatClient(base_url="http://localhost:8000") as client:
            await client.login("username", "password")
            async for chunk in client.chat_stream("Hello!"):
                print(chunk.content, end="", flush=True)
    """

    def __init__(
        self,
        base_url: str = "http://localhost:8000",
        timeout: float = 30.0,
        headers: Optional[Dict[str, str]] = None,
        max_retries: int = 3,
        limits: Optional[httpx.Limits] = None,
    ):
        """
        Initialize async client.

        Args:
            base_url: API base URL
            timeout: Request timeout in seconds
            headers: Default headers
            max_retries: Maximum retry attempts
            limits: HTTPX limits for connection pooling
        """
        self.base_url = base_url.rstrip("/")
        self.timeout = timeout
        self.max_retries = max_retries

        default_headers = {
            "Content-Type": "application/json",
            "Accept": "application/json",
        }
        if headers:
            default_headers.update(headers)

        if limits is None:
            limits = httpx.Limits(
                max_keepalive_connections=20,
                max_connections=100,
                keepalive_expiry=30.0,
            )

        self._client = httpx.AsyncClient(
            base_url=self.base_url,
            timeout=timeout,
            headers=default_headers,
            limits=limits,
        )

        self._token: Optional[str] = None
        self._refresh_token: Optional[str] = None

    async def __aenter__(self):
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb):
        await self.close()

    @property
    def token(self) -> Optional[str]:
        return self._token

    @property
    def refresh_token(self) -> Optional[str]:
        return self._refresh_token

    def set_token(self, access_token: str, refresh_token: Optional[str] = None) -> None:
        """Set authentication tokens."""
        self._token = access_token
        if refresh_token:
            self._refresh_token = refresh_token
        self._client.headers["Authorization"] = f"Bearer {access_token}"

    def clear_tokens(self) -> None:
        """Clear authentication tokens."""
        self._token = None
        self._refresh_token = None
        self._client.headers.pop("Authorization", None)

    async def close(self) -> None:
        """Close HTTP client."""
        await self._client.aclose()

    async def _request(
        self,
        method: str,
        path: str,
        *,
        json: Optional[Dict[str, Any]] = None,
        params: Optional[Dict[str, Any]] = None,
        headers: Optional[Dict[str, str]] = None,
        retries: Optional[int] = None,
        stream: bool = False,
    ) -> Union[httpx.Response, httpx.Response]:
        """Make HTTP request with retry logic."""
        url = f"/api/v1{path}" if not path.startswith("/api/v1") else path
        retries = retries if retries is not None else self.max_retries

        @standard_retry
        async def _do_request():
            response = await self._client.request(
                method=method,
                url=url,
                json=json,
                params=params,
                headers=headers,
                stream=stream,
            )

            # Handle rate limit
            if response.status_code == 429:
                retry_after = response.headers.get("Retry-After", "60")
                raise RateLimitError(
                    "Rate limit exceeded",
                    status_code=429,
                    retry_after=int(retry_after),
                )

            # Handle other errors
            if response.status_code >= 400:
                error_data = None
                try:
                    error_data = response.json()
                except Exception:
                    error_data = {"detail": response.text or f"HTTP {response.status_code}"}

                error = map_http_error(
                    response.status_code,
                    error_data.get("detail", f"HTTP {response.status_code}"),
                    error_data,
                )
                raise error

            return response

        return await _do_request()

    async def _get(self, path: str, **kwargs) -> Dict[str, Any]:
        response = await self._request("GET", path, **kwargs)
        return response.json()

    async def _post(self, path: str, **kwargs) -> Dict[str, Any]:
        response = await self._request("POST", path, **kwargs)
        return response.json()

    async def _put(self, path: str, **kwargs) -> Dict[str, Any]:
        response = await self._request("PUT", path, **kwargs)
        return response.json()

    async def _delete(self, path: str, **kwargs) -> Dict[str, Any]:
        response = await self._request("DELETE", path, **kwargs)
        if response.status_code == 204:
            return {}
        return response.json()

    # ==================== Authentication ====================

    async def login(self, username: str, password: str) -> TokenResponse:
        """
        Authenticate user and get tokens.

        Args:
            username: Username
            password: Password

        Returns:
            TokenResponse with access_token and refresh_token
        """
        data = await self._post("/auth/login", json={"username": username, "password": password})
        token_response = TokenResponse(**data)
        self.set_token(token_response.access_token, token_response.refresh_token)
        return token_response

    async def refresh_token(self, refresh_token: Optional[str] = None) -> TokenResponse:
        """
        Refresh access token.

        Args:
            refresh_token: Refresh token (uses stored if not provided)

        Returns:
            New TokenResponse
        """
        token = refresh_token or self._refresh_token
        if not token:
            raise AuthenticationError("No refresh token available")

        data = await self._post("/auth/refresh", json={"refresh_token": token})
        token_response = TokenResponse(**data)
        self.set_token(token_response.access_token, token_response.refresh_token)
        return token_response

    async def logout(self, refresh_token: Optional[str] = None) -> Dict[str, str]:
        """
        Logout by revoking refresh token.

        Args:
            refresh_token: Refresh token to revoke (uses stored if not provided)

        Returns:
            Success message
        """
        token = refresh_token or self._refresh_token
        if token:
            await self._post("/auth/logout", json={"refresh_token": token})
        self.clear_tokens()
        return {"message": "Logged out"}

    # ==================== API Keys ====================

    async def create_api_key(self, name: str, expires_in_days: Optional[int] = None) -> APIKeyResponse:
        """
        Create a new API key.

        Args:
            name: Key name
            expires_in_days: Optional expiry in days

        Returns:
            APIKeyResponse with full key (only returned once)
        """
        data = await self._post("/auth/api-keys", json={
            "name": name,
            "expires_in_days": expires_in_days,
        })
        return APIKeyResponse(**data)

    async def list_api_keys(self) -> List[APIKeyResponse]:
        """List all API keys for current user."""
        data = await self._get("/auth/api-keys")
        return [APIKeyResponse(**item) for item in data]

    async def revoke_api_key(self, prefix: str) -> Dict[str, str]:
        """
        Revoke an API key by prefix.

        Args:
            prefix: API key prefix (e.g., jebat_abc123)
        """
        await self._delete(f"/auth/api-keys/{prefix}")
        return {"message": "API key revoked"}

    # ==================== User Profile ====================

    async def get_profile(self) -> UserResponse:
        """Get current user profile."""
        data = await self._get("/auth/me")
        return UserResponse(**data)

    async def get_user(self, user_id: str) -> UserResponse:
        """Get user by ID (admin only)."""
        data = await self._get(f"/auth/users/{user_id}")
        return UserResponse(**data)

    # ==================== Chat ====================

    async def chat(
        self,
        message: str,
        user_id: Optional[str] = None,
        mode: str = "deliberate",
        timeout: int = 30,
    ) -> ChatResponse:
        """
        Send a chat message (blocking).

        Args:
            message: User message
            user_id: User ID (uses current user if not provided)
            mode: Thinking mode (fast, deliberate, deep, strategic, creative, critical)
            timeout: Request timeout in seconds

        Returns:
            ChatResponse
        """
        data = await self._post("/chat", json={
            "message": message,
            "user_id": user_id,
            "mode": mode,
            "timeout": timeout,
        })
        return ChatResponse(**data)

    async def chat_stream(
        self,
        message: str,
        user_id: Optional[str] = None,
        mode: str = "deliberate",
        timeout: int = 30,
    ) -> AsyncIterator[StreamChunk]:
        """
        Stream chat response via Server-Sent Events.

        Args:
            message: User message
            user_id: User ID
            mode: Thinking mode
            timeout: Request timeout

        Yields:
            StreamChunk objects
        """
        url = f"{self.base_url}/api/v1/chat"
        payload = {
            "message": message,
            "user_id": user_id,
            "mode": mode,
            "timeout": timeout,
            "stream": True,
        }

        async with self._client.stream("POST", url, json=payload, timeout=timeout) as response:
            async for line in response.aiter_lines():
                if line.startswith("data: "):
                    data = line[6:].strip()
                    if data == "[DONE]":
                        break
                    try:
                        yield StreamChunk(
                            type="content",
                            content=data,
                        )
                    except Exception as e:
                        yield StreamChunk(type="error", content=str(e))

    async def chat_websocket(
        self,
        message: str,
        mode: str = "deliberate",
        user_id: Optional[str] = None,
    ) -> AsyncIterator[StreamChunk]:
        """
        Stream chat via WebSocket (requires separate WebSocket connection).

        Args:
            message: User message
            mode: Thinking mode
            user_id: User ID

        Yields:
            StreamChunk objects
        """
        from .websocket import JebatWebSocketClient

        ws_url = self.base_url.replace("http", "ws") + "/ws/chat"
        async with JebatWebSocketClient(ws_url, self._token) as ws:
            await ws.send_chat(message, mode)
            async for chunk in ws.stream({"message": message, "mode": mode}):
                yield chunk

    # ==================== Memories ====================

    async def list_memories(
        self,
        user_id: Optional[str] = None,
        query: Optional[str] = None,
        layer: Optional[str] = None,
        limit: int = 20,
        offset: int = 0,
    ) -> MemoryListResponse:
        """List memories with optional filtering."""
        params = {"limit": limit, "offset": offset}
        if user_id:
            params["user_id"] = user_id
        if query:
            params["query"] = query
        if layer:
            params["layer"] = layer

        data = await self._get("/memories", params=params)
        return MemoryListResponse(**data)

    async def create_memory(
        self,
        content: str,
        user_id: str,
        layer: str = "M1_EPISODIC",
        metadata: Optional[Dict[str, Any]] = None,
    ) -> MemoryItem:
        """Create a new memory."""
        data = await self._post("/memories", json={
            "content": content,
            "user_id": user_id,
            "layer": layer,
            "metadata": metadata,
        })
        return MemoryItem(**data)

    async def search_memories(
        self,
        query: str,
        user_id: Optional[str] = None,
        layer: Optional[str] = None,
        limit: int = 20,
        offset: int = 0,
    ) -> MemoryListResponse:
        """Search memories."""
        params = {"query": query, "limit": limit, "offset": offset}
        if user_id:
            params["user_id"] = user_id
        if layer:
            params["layer"] = layer

        data = await self._get("/memories", params=params)
        return MemoryListResponse(**data)

    # ==================== Agents ====================

    async def list_agents(self) -> AgentListResponse:
        """List all agents."""
        data = await self._get("/agents")
        return AgentListResponse(**data)

    async def get_agent(self, agent_id: str) -> AgentInfo:
        """Get agent by ID."""
        data = await self._get(f"/agents/{agent_id}")
        return AgentInfo(**data)

    async def execute_agent(
        self,
        task: str,
        agent_id: Optional[str] = None,
        mode: str = "deliberate",
        timeout: int = 30,
    ) -> AgentExecuteResponse:
        """Execute an agent task."""
        data = await self._post("/agents/execute", json={
            "task": task,
            "agent_id": agent_id,
            "mode": mode,
            "timeout": timeout,
        })
        return AgentExecuteResponse(**data)

    # ==================== Channels ====================

    async def list_channels(self) -> List[ChannelInfo]:
        """List all channel configurations."""
        data = await self._get("/channels")
        return [ChannelInfo(**item) for item in data]

    async def get_channel(self, channel_id: str) -> ChannelInfo:
        """Get channel by ID."""
        data = await self._get(f"/channels/{channel_id}")
        return ChannelInfo(**data)

    async def update_channel_config(
        self,
        channel_id: str,
        config: Dict[str, Any],
    ) -> ChannelInfo:
        """Update channel configuration."""
        data = await self._put(f"/channels/{channel_id}", json={"config": config})
        return ChannelInfo(**data)

    async def connect_channel(self, channel_id: str) -> Dict[str, str]:
        """Connect a channel."""
        data = await self._post(f"/channels/{channel_id}/connect")
        return data

    async def disconnect_channel(self, channel_id: str) -> Dict[str, str]:
        """Disconnect a channel."""
        data = await self._post(f"/channels/{channel_id}/disconnect")
        return data

    # ==================== Monitoring ====================

    async def get_health(self) -> HealthResponse:
        """Get system health."""
        data = await self._get("/health")
        return HealthResponse(**data)

    async def get_status(self) -> StatusResponse:
        """Get system status."""
        data = await self._get("/status")
        return StatusResponse(**data)

    async def get_metrics(self) -> MetricsResponse:
        """Get system metrics."""
        data = await self._get("/metrics")
        return MetricsResponse(**data)

    # ==================== Swarm ====================

    async def execute_swarm(
        self,
        description: str,
        user_id: Optional[str] = None,
        execution_mode: str = "consensus",
        required_capabilities: Optional[List[str]] = None,
        enable_search: bool = True,
        search_queries: Optional[List[str]] = None,
        max_agents: int = 5,
    ) -> SwarmTaskResponse:
        """Execute a task through the swarm orchestrator."""
        data = await self._post("/swarm/execute", json={
            "description": description,
            "user_id": user_id,
            "execution_mode": execution_mode,
            "required_capabilities": required_capabilities,
            "enable_search": enable_search,
            "search_queries": search_queries,
            "max_agents": max_agents,
        })
        return SwarmTaskResponse(**data)

    async def plan_swarm(
        self,
        description: str,
        execution_mode: str = "consensus",
        required_capabilities: Optional[List[str]] = None,
        max_agents: int = 5,
    ) -> SwarmPlanResponse:
        """Get swarm execution plan without executing."""
        data = await self._post("/swarm/plan", json={
            "description": description,
            "execution_mode": execution_mode,
            "required_capabilities": required_capabilities,
            "max_agents": max_agents,
        })
        return SwarmPlanResponse(**data)

    # ==================== Context Manager ====================

    @asynccontextmanager
    async def session(self):
        """Context manager for temporary session overrides."""
        old_token = self._token
        old_refresh = self._refresh_token
        try:
            yield self
        finally:
            self._token = old_token
            self._refresh_token = old_refresh
            if old_token:
                self._client.headers["Authorization"] = f"Bearer {old_token}"
            else:
                self._client.headers.pop("Authorization", None)

File: jebat-sdk-ts/src/async-client.ts
/**
 * Asynchronous JEBAT API client with full streaming support.
 */

import { HttpClient, type RequestOptions } from './http-client';
import { JebatWebSocketClient, type StreamChunk as WSStreamChunk } from './websocket';
import type { ClientOptions, ThinkingMode, MemoryLayer, SwarmExecutionMode } from './types';
import type {
  // Auth
  TokenRequest, TokenResponse, RefreshTokenRequest,
  APIKeyCreateRequest, APIKeyResponse, UserResponse,
  // Chat
  ChatRequest, ChatResponse, Message, OpenAIChatRequest, OpenAIChatResponse,
  // Memories
  MemoryItem, MemoryListResponse, MemoryCreateRequest, MemorySearchRequest,
  // Agents
  AgentInfo, AgentListResponse, AgentExecuteRequest, AgentExecuteResponse,
  // Channels
  ChannelInfo, ChannelConfig, ChannelListResponse,
  // Monitoring
  HealthResponse, StatusResponse, MetricsResponse,
  // Swarm
  SwarmTaskRequest, SwarmTaskResponse, SwarmPlanResponse,
} from './models';

/**
 * Stream chunk for SSE streaming.
 */
export interface StreamChunk {
  type: 'content' | 'thinking' | 'complete' | 'error';
  content: string;
  data?: unknown;
}

/**
 * Async JEBAT API client with full streaming support.
 *
 * @example
 * ```typescript
 * const client = new AsyncJebatClient({ baseUrl: 'http://localhost:8000' });
 * await client.login('username', 'password');
 *
 * // Streaming chat
 * for await (const chunk of client.chatStream('Hello!')) {
 *   process.stdout.write(chunk.content);
 * }
 *
 * // Parallel requests
 * const [health, agents] = await Promise.all([
 *   client.getHealth(),
 *   client.listAgents(),
 * ]);
 * ```
 */
export class AsyncJebatClient {
  private http: HttpClient;
  private wsClient: JebatWebSocketClient | null = null;

  constructor(options: ClientOptions) {
    this.http = new HttpClient(options);
  }

  // ==================== Delegated Properties ====================

  get baseUrl(): string {
    return (this.http as any).baseUrl;
  }

  get token(): string | null {
    return this.http.getToken();
  }

  get refreshToken(): string | null {
    return this.http.getRefreshToken();
  }

  setToken(accessToken: string, refreshToken?: string): void {
    this.http.setToken(accessToken, refreshToken);
  }

  clearTokens(): void {
    this.http.clearTokens();
  }

  isAuthenticated(): boolean {
    return this.http.isAuthenticated();
  }

  // ==================== Context Manager ====================

  async [Symbol.asyncDispose](): Promise<void> {
    await this.close();
  }

  async close(): Promise<void> {
    if (this.wsClient) {
      await this.wsClient.disconnect();
      this.wsClient = null;
    }
  }

  // ==================== Authentication ====================

  /**
   * Authenticate user and get tokens.
   */
  async login(username: string, password: string): Promise<TokenResponse> {
    const response = await this.http.post<TokenResponse>('/api/v1/auth/login', { username, password });
    this.http.setToken(response.access_token, response.refresh_token);
    return response;
  }

  /**
   * Refresh access token.
   */
  async refreshAccessToken(refreshToken?: string): Promise<TokenResponse> {
    const token = refreshToken ?? this.http.getRefreshToken();
    if (!token) {
      throw new Error('No refresh token available');
    }
    const response = await this.http.post<TokenResponse>('/api/v1/auth/refresh', { refresh_token: token });
    this.http.setToken(response.access_token, response.refresh_token);
    return response;
  }

  /**
   * Logout by revoking refresh token.
   */
  async logout(refreshToken?: string): Promise<{ message: string }> {
    const token = refreshToken ?? this.http.getRefreshToken();
    if (token) {
      await this.http.post('/api/v1/auth/logout', { refresh_token: token });
    }
    this.http.clearTokens();
    return { message: 'Logged out' };
  }

  // ==================== API Keys ====================

  async createApiKey(name: string, expiresInDays?: number): Promise<APIKeyResponse> {
    return this.http.post<APIKeyResponse>('/api/v1/auth/api-keys', { name, expires_in_days: expiresInDays });
  }

  async listApiKeys(): Promise<APIKeyResponse[]> {
    return this.http.get<APIKeyResponse[]>('/api/v1/auth/api-keys');
  }

  async revokeApiKey(prefix: string): Promise<{ message: string }> {
    await this.http.delete(`/api/v1/auth/api-keys/${prefix}`);
    return { message: 'API key revoked' };
  }

  // ==================== User Profile ====================

  async getProfile(): Promise<UserResponse> {
    return this.http.get<UserResponse>('/api/v1/auth/me');
  }

  async getUser(userId: string): Promise<UserResponse> {
    return this.http.get<UserResponse>(`/api/v1/auth/users/${userId}`);
  }

  // ==================== Chat ====================

  /**
   * Send a chat message (blocking).
   */
  async chat(
    message: string,
    options: {
      userId?: string;
      mode?: ThinkingMode;
      timeout?: number;
    } = {}
  ): Promise<ChatResponse> {
    return this.http.post<ChatResponse>('/api/v1/chat', {
      message,
      user_id: options.userId,
      mode: options.mode ?? 'deliberate',
      timeout: options.timeout ?? 30,
    });
  }

  /**
   * Stream chat response via Server-Sent Events.
   */
  async *chatStream(
    message: string,
    options: {
      userId?: string;
      mode?: ThinkingMode;
      timeout?: number;
      signal?: AbortSignal;
    } = {}
  ): AsyncGenerator<StreamChunk, void, unknown> {
    const url = new URL('/api/v1/chat', this.baseUrl);
    const payload = {
      message,
      user_id: options.userId,
      mode: options.mode ?? 'deliberate',
      timeout: options.timeout ?? 30,
      stream: true,
    };

    const response = await fetch(url.toString(), {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Accept': 'text/event-stream',
        ...(this.token ? { 'Authorization': `Bearer ${this.token}` } : {}),
      },
      body: JSON.stringify(payload),
      signal: options.signal,
    });

    if (!response.ok) {
      const error = await response.json().catch(() => ({ detail: `HTTP ${response.status}` }));
      throw new Error(error.detail || `HTTP ${response.status}`);
    }

    const reader = response.body?.getReader();
    if (!reader) {
      throw new Error('No response body');
    }

    const decoder = new TextDecoder();
    let buffer = '';

    try {
      while (true) {
        const { done, value } = await reader.read();
        if (done) break;

        buffer += decoder.decode(value, { stream: true });
        const lines = buffer.split('\n');
        buffer = lines.pop() || '';

        for (const line of lines) {
          if (line.startsWith('data: ')) {
            const data = line.slice(6).trim();
            if (data === '[DONE]') {
              yield { type: 'complete', content: '' };
              return;
            }
            yield { type: 'content', content: data };
          }
        }
      }
    } finally {
      reader.releaseLock();
    }
  }

  /**
   * Stream chat via WebSocket (lower latency, bidirectional).
   */
  async connectWebSocket(): Promise<JebatWebSocketClient> {
    if (this.wsClient?.isConnected) {
      return this.wsClient;
    }

    const wsUrl = this.baseUrl.replace('http', 'ws') + '/ws/chat';
    this.wsClient = new JebatWebSocketClient(wsUrl, this.token!);
    await this.wsClient.connect();
    return this.wsClient;
  }

  /**
   * Stream chat using WebSocket.
   */
  async *chatWebSocket(
    message: string,
    mode: ThinkingMode = 'deliberate'
  ): AsyncGenerator<WSStreamChunk, void, unknown> {
    const ws = await this.connectWebSocket();
    yield* ws.stream({ message, mode });
  }

  // ==================== Memories ====================

  async listMemories(options: {
    userId?: string;
    query?: string;
    layer?: MemoryLayer;
    limit?: number;
    offset?: number;
  } = {}): Promise<MemoryListResponse> {
    const params: Record<string, unknown> = {
      limit: options.limit ?? 20,
      offset: options.offset ?? 0,
    };
    if (options.userId) params['user_id'] = options.userId;
    if (options.query) params['query'] = options.query;
    if (options.layer) params['layer'] = options.layer;

    return this.http.get<MemoryListResponse>('/api/v1/memories', params);
  }

  async createMemory(
    content: string,
    userId: string,
    layer: MemoryLayer = 'M1_EPISODIC',
    metadata?: Record<string, unknown>
  ): Promise<MemoryItem> {
    return this.http.post<MemoryItem>('/api/v1/memories', {
      content,
      user_id: userId,
      layer,
      metadata,
    });
  }

  async searchMemories(options: MemorySearchRequest): Promise<MemoryListResponse> {
    return this.listMemories(options);
  }

  // ==================== Agents ====================

  async listAgents(): Promise<AgentListResponse> {
    return this.http.get<AgentListResponse>('/api/v1/agents');
  }

  async getAgent(agentId: string): Promise<AgentInfo> {
    return this.http.get<AgentInfo>(`/api/v1/agents/${agentId}`);
  }

  async executeAgent(task: string, options: {
    agentId?: string;
    mode?: ThinkingMode;
    timeout?: number;
  } = {}): Promise<AgentExecuteResponse> {
    return this.http.post<AgentExecuteResponse>('/api/v1/agents/execute', {
      task,
      agent_id: options.agentId,
      mode: options.mode ?? 'deliberate',
      timeout: options.timeout ?? 30,
    });
  }

  // ==================== Channels ====================

  async listChannels(): Promise<ChannelInfo[]> {
    return this.http.get<ChannelInfo[]>('/api/v1/channels');
  }

  async getChannel(channelId: string): Promise<ChannelInfo> {
    return this.http.get<ChannelInfo>(`/api/v1/channels/${channelId}`);
  }

  async updateChannelConfig(channelId: string, config: Record<string, unknown>): Promise<ChannelInfo> {
    return this.http.put<ChannelInfo>(`/api/v1/channels/${channelId}`, { config });
  }

  async connectChannel(channelId: string): Promise<{ message: string }> {
    return this.http.post(`/api/v1/channels/${channelId}/connect`);
  }

  async disconnectChannel(channelId: string): Promise<{ message: string }> {
    return this.http.post(`/api/v1/channels/${channelId}/disconnect`);
  }

  // ==================== Monitoring ====================

  async getHealth(): Promise<HealthResponse> {
    return this.http.get<HealthResponse>('/api/v1/health');
  }

  async getStatus(): Promise<StatusResponse> {
    return this.http.get<StatusResponse>('/api/v1/status');
  }

  async getMetrics(): Promise<MetricsResponse> {
    return this.http.get<MetricsResponse>('/api/v1/metrics');
  }

  // ==================== Swarm ====================

  async executeSwarm(
    description: string,
    options: {
      userId?: string;
      executionMode?: SwarmExecutionMode;
      requiredCapabilities?: string[];
      enableSearch?: boolean;
      searchQueries?: string[];
      maxAgents?: number;
    } = {}
  ): Promise<SwarmTaskResponse> {
    return this.http.post<SwarmTaskResponse>('/api/v1/swarm/execute', {
      description,
      user_id: options.userId,
      execution_mode: options.executionMode ?? 'consensus',
      required_capabilities: options.requiredCapabilities,
      enable_search: options.enableSearch ?? true,
      search_queries: options.searchQueries,
      max_agents: options.maxAgents ?? 5,
    });
  }

  async planSwarm(
    description: string,
    options: {
      executionMode?: SwarmExecutionMode;
      requiredCapabilities?: string[];
      maxAgents?: number;
    } = {}
  ): Promise<SwarmPlanResponse> {
    return this.http.post<SwarmPlanResponse>('/api/v1/swarm/plan', {
      description,
      execution_mode: options.executionMode ?? 'consensus',
      required_capabilities: options.requiredCapabilities,
      max_agents: options.maxAgents ?? 5,
    });
  }

  // ==================== Utility ====================

  /**
   * Execute multiple requests in parallel.
   */
  async parallel<T extends Promise<unknown>[]>(
    requests: { [K in keyof T]: () => T[K] }
  ): Promise<{ [K in keyof T]: Awaited<T[K]> }> {
    return Promise.all(requests.map(fn => fn())) as Promise<{ [K in keyof T]: Awaited<T[K]> }>;
  }
}